compiler/rustc_metadata/src/locator.rs

```diff
@@ -723,18 +733,49 @@ impl<'a> CrateLocator<'a> {
             rejected_via_kind: self.rejected_via_kind,
             rejected_via_version: self.rejected_via_version,
             rejected_via_filename: self.rejected_via_filename,
+            rejected_via_invalid: self.rejected_via_invalid,
         })
     }
 }
 
+enum MetadataError<'a> {
+    /// The file was missing.
+    NotPresent(&'a Path),
+    /// The file was present and valid, but built by a different and incompatible version of the compiler.
+    WrongVersion(&'a Path),
+    /// The file was present and invalid.
+    LoadFailure(String),
+}
+
+impl fmt::Display for MetadataError<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NotPresent(filename) => {
+                f.write_str(&format!("no such file: '{}'", filename.display()))
+            }
+            Self::WrongVersion(filename) => f.write_str(&format!(
+                "incompatible metadata version found: '{}'",
+                filename.display()
+            )),
+            Self::LoadFailure(msg) => f.write_str(msg),
```

Comment:
```suggestion
            MetadataError::LoadFailure(msg) => f.write_str(msg),
```
`Self` in non-generic context almost always hurts code searchability and readability.